CCO-20:TP docs for AWS STS (#29545)
openshift/openshift-docs: authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc

> [id="cco-mode-sts"]

Author: :sparkles: NEW STS content here :eyes:

= Using STS mode
include::modules/common-attributes.adoc[]
:context: cco-mode-sts
toc::[]

> STS mode is supported for Amazon Web Services (AWS).

Reviewer: not yet supported :)

Author: :detective:  changed to say TP for AWS

:FeatureName: Support for AWS Secure Token Service (STS)
include::modules/technology-preview.adoc[leveloffset=+0]

> In STS mode, the Cloud Credential Operator (CCO) uses AWS Secure Token Service (STS) to assign components IAM roles that provide short-term, limited-privilege security credentials.

Reviewer: ```suggestion
In STS mode, the individual OpenShift cluster components use AWS Secure Token Service (STS) to acquire short-term, limited-privilege security credentials associated with IAM Roles that are specific to each component making AWS API calls.
```

> Requests for new and refreshed credentials are automated by using an AWS IAM OpenID Connect (OIDC) identity provider. {product-title} signs service account tokens that are trusted by AWS IAM, and can be projected into a pod and used for authentication. Tokens are refreshed after one hour.

Reviewer: ```suggestion
Requests for new and refreshed credentials are automated by using an appropriately configured AWS IAM OpenID Connect (OIDC) identity provider combined with AWS IAM Roles. {product-title} signs service account tokens that are trusted by AWS IAM, and can be projected into a pod and used for authentication. Tokens are refreshed after one hour.
```

.STS authentication flow
image::sts_flow.png[STS authentication flow]

> The implementation of STS mode changes the format of the AWS credentials root secret in the `kube-system` namespace.

Reviewer: ```suggestion
Running with STS changes the content of the AWS credentials provided to individual OpenShift components.
```

There is no root secret in `kubs-system` when running with STS

.AWS secret format using a mode other than STS

[source,yaml]
----
apiVersion: v1
kind: Secret

> metadata:
>   namespace: kube-system
>   name: aws-creds
> data:
>   aws_access_key_id: <base64-encoded-access-key-id>
>   aws_secret_access_key: <base64-encoded-secret-access-key>

Reviewer: ```suggestion
metadata:
  namespace: target-namespace
  name: target-secret-name
data:
  aws_access_key_id: <base64-encoded-access-key-id>
  aws_secret_access_key: <base64-encoded-secret-access-key>
```

----

.AWS secret format using STS mode

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:

>   namespace: kube-system
>   name: aws-creds
> data:
>   role_name: <operator-role-name> <1>
>   web_identity_token_file: <path-to-token> <2>

Reviewer: ```suggestion
  namespace: target-namespace
  name: target-secret-name
data:
  role_name: <operator-role-name> <1>
  web_identity_token_file: <path-to-token> <2>
```

----
<1> The IAM role for the cluster
+
<needinfo> Is this accurate/informative? or is there a better way to describe it?

> <2> The path to the service account tokens.

Reviewer: ```suggestion
<2> The path to the service account token inside the pod. /var/run/secrets/openshift/serviceaccount/token by convention for OpenShift components.
```

[id="sts-mode-installing"]
== Installing an {product-title} cluster configured for STS mode

In {product-title} version 4.7, installing a cluster that is configured to use the CCO in STS mode requires the following steps:

. xref:#sts-mode-installing-preparing[Preparing the installer]
. Creating AWS resources xref:#sts-mode-installing-config-preflight[using the `sts-preflight` tool] or xref:#sts-mode-installing-config-manual[manually]
. xref:#sts-mode-installing-running-installer[Running the installer]
. xref:#sts-mode-installing-verifying[Verifying the installation]

[NOTE]
====
<needinfo>

> Are they basically following the link:https://docs.openshift.com/container-platform/4.6/installing/installing_aws/installing-aws-customizations.html#installation-initializing_installing-aws-customizations[Installing a cluster on AWS with customizations] procedure here, with some specific differences for STS (and of course the addition of the preflight stuff)?

Reviewer: It's a slightly more advanced version of that documentation. B/c we need to also slip in custom manifests files before launching the "real" installation.

====

[id="sts-mode-installing-preparing"]
=== Preparing the installer

.Prerequisites
* Obtain the pull secret for your cluster.

.Procedure

. Set the `$RELEASE_IMAGE` environment variable to use a sufficiently new {product-title} release:
+
[source,terminal]
----

> <needinfo>

Reviewer: "Sufficiently new" was back when there were still missing pieces that hadn't merged into 4.7. Now that 4.7 has all the needed PRs, you only need OpenShift version 4.7.

> - Need information about how to do this (commands, output if applicable)

Reviewer: Once you have the openshfit-install binary, you can run `openshift-install version` and it will show you the release image associated with the specific version of the installer that is being run.
It comes out in a non-human-friendly format though (it doesn't actually encode the version string into the release image URL/text:

```
[jdiaz@minigoomba os-install-4.6.9]$ ./openshift-install version
./openshift-install 4.6.9
built from commit a48ad4a15b42102d1747d2f5f3b635deffb950b5
release image quay.io/openshift-release-dev/ocp-release@sha256:43d5c84169a4b3ff307c29d7374f6d69a707de15e9fa90ad352b432f77c0cead
```

But if you browse to https://quay.io/repository/openshift-release-dev/ocp-release?tab=tags and search/filter for `4.6.9-x86_64` you will see that the SHA256 for the 4.6.9-x86_64 matches the release image printed out by the installer.

![image](https://user-images.githubusercontent.com/4999586/108235538-e7ab1b80-7113-11eb-918e-c3bd6bac3749.png)

Author: I _think_ I resolved the bit with these first steps by pulling from the manually creating IAM content. Hopefully refreshed PR will get this cleaned up.

- Is it this step? https://docs.openshift.com/container-platform/4.6/installing/installing_bare_metal_ipi/ipi-install-installation-workflow.html#retrieving-the-openshift-installer_ipi-install-installation-workflow
- Will also need to know more specifically about "sufficiently new".
----

. Extract the AWS `CredentialsRequest` objects from the release image:
+
[source,terminal]
----
$ mkdir credreqs ; oc adm release extract --cloud=aws --credentials-requests $RELEASE_IMAGE --to=./credreqs ; cat ./credreqs/*.yaml > credreqs.yaml
----

. Extract the {product-title} installer from the release image:
+
[source,terminal]
----
$ oc adm release extract --command=openshift-install $RELEASE_IMAGE

> ----

Reviewer: Again, this is a step from back before all the PRs had merged. This shouldn't be necessary, as the user can just download the installer from whatever regular location they normally get it from.

. Create the `install-config.yaml` file:
+
[source,terminal]
----
$ ./openshift-install create install-config
----

. Configure the cluster to install with the CCO in manual mode:
+
[source,terminal]
----
$ echo "credentialsMode: Manual" >> install-config.yaml
----

. Create install manifests:
+
[source,terminal]
----
$ ./openshift-install create manifests
----

[id="sts-mode-installing-config-preflight"]
=== Creating AWS resources using the sts-preflight tool

When installing an {product-title} cluster that is configured to use the CCO in STS mode, you can use the `sts-preflight` tool to create the required AWS resources. This tool, which is hosted in an external repository, is provided to help configure a cluster that uses STS mode while the feature is a link:https://access.redhat.com/support/offerings/techpreview[Technology Preview].

As an alternative, you can xref:#sts-mode-installing-config-manual[create the required AWS resources manually].

.Prerequisites
* Download the link:https://github.com/sjenning/sts-preflight[`sts-preflight` tool].
* Ensure that you have an AWS account with sufficient permissions.
+

> <needinfo> what are these? the same permissions we get users to check in xref:../../installing/installing_aws/manually-creating-iam.adoc#manually-creating-iam-aws[Manually Creating IAM for AWS]?

Reviewer: We haven't compiled that precise list of permissions yet, which is why we have the vague "sufficient permissions" in the text :/

Author: ack

.Procedure

. Run the following command to:
+
--
** Generate public and private `ServiceAccount` object signing keys.
** Create an S3 bucket.
** Upload the OIDC config into the S3 bucket.
** Set up an IAM Identity Provider that trusts the S3 bucket.
** Create an IAM role with administrator rights to use for installation.
** Create IAM roles for each AWS `CredentialsRequest` object.
+
[source,terminal]
----
$ ./sts-preflight create --infra-name <aws-infra-name> --region <aws-region> --credentials-requests-to-roles <path-to-credreqs.yaml>

> <needinfo> if any of these variables are not self-explanatory, we need to define what they are. I think region and path are obvious, but will the user know what to put for `aws-infra-name`?

Reviewer: aws-infra-name is the just a prefix used when the sts-preflight tool is naming the objects that will be created in AWS.

+
[source,json]
----
"Condition": {
  "StringEquals": {
    "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [
      "system:serviceaccount:openshift-machine-api:machine-api-controllers"
    ]
  }
}
----

. Copy the manifests that the `sts-preflight` tool created to the `manifests` directory in the same location as the `install-config.yaml` file:
+
[source,terminal]
----
$ cp _output/manifests/* <path-to-install-config.yaml-dir>/manifests/
----

. Copy the private key for the `ServiceAccount` object signer to the same location as the `install-config.yaml` file:
+
[source,terminal]
----
$ cp -a _output/tls <path-to-install-config.yaml-dir>
----

. Create a token that is trusted by the OIDC identity provider:
+
[source,terminal]
----
$ ./sts-preflight token
----
+
This token is used to assume a role to get the credentials that are required to start the installation. The token is created in the `_output/token` directory.

. Use the token to assume the installer role and mint the credentials that are required for the {product-title} installation process.
+
[source,terminal]
----
$ ./sts-preflight assume
----
+

> <needinfo> it would be helpful to have an example output here since the next step refers to it.

Reviewer: ```
[jdiaz@minigoomba sts-preflight (master %=)]$ ./sts-preflight assume
Run these commands to use the STS credentials
export AWS_ACCESS_KEY_ID=accessKeyId
export AWS_SECRET_ACCESS_KEY=secretAccessKey
export AWS_SESSION_TOKEN=superLongObnoxiousAwsSessionTokenHere
```

Author: ah - so basically self-documented. Never mind then :)

For example for the openshift-machine-api, the Secret file would look like:

$ cat manifests/openshift-machine-api-aws-cloud-credentials-credentials.yaml
apiVersion: v1
stringData:
  credentials: |-
    [default]
    role_arn = arn:aws:iam::123456789012:role/openshift-machine-api-aws-cloud-credentials
    web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
kind: Secret
metadata:
  name: aws-cloud-credentials
  namespace: openshift-machine-api
type: Opaque

Run the installer:
$ ./openshift-install create cluster
////

[id="sts-mode-installing-running-installer"]
=== Running the installer

. Run the {product-title} installer
+
[source,terminal]
----
$ ./openshift-install create cluster --log-level=debug
----

[id="sts-mode-installing-verifying"]
=== Verifying the installation

. Connect to the {product-title} cluster.

. Verify that the cluster does not have `root` credentials:
+
[source,terminal]
----
$ oc get secrets -n kube-system aws-creds
----
+
The output should look similar to:
+
[source,terminal]
----
Error from server (NotFound): secrets "aws-creds" not found

> <needinfo> I ran that on the Katacoda OCP playground but don't entirely trust the formatting of what I got back - is that correct for a real cluster?

Reviewer: That output looks correct.

[id="cco-mode-sts"]

> = Using manual mode with STS

Reviewer 2: Do we need CCO in the heading, or the intro paragraphs?

include::modules/common-attributes.adoc[]
:context: cco-mode-sts
toc::[]

> Manual mode with STS is available as a link:https://access.redhat.com/support/offerings/techpreview[Technology Preview] for Amazon Web Services (AWS).

Reviewer 2: It seems you modeled the existing topics. Not sure we should lead with this paragraph, rather describe what each mode is then state that it's TP. If I don't know what the mode is, why do I care if it's supported? Just a thought.

Author: I'm worried about not having the supported platforms up front on these, because I would hate for someone using e.g. RHV to have to read all about mint mode and then find out they can't use it. This particular page is I guess a bit odd in that we can't just say it's supported on AWS, since it's not supported. But that would change to the more normal statement when this is GA.

:FeatureName: Support for AWS Secure Token Service (STS)
include::modules/technology-preview.adoc[leveloffset=+0]

In manual mode with STS, the individual {product-title} cluster components use AWS Secure Token Service (STS) to assign components IAM roles that provide short-term, limited-privilege security credentials. These credentials are associated with IAM roles that are specific to each component that makes AWS API calls.

Requests for new and refreshed credentials are automated by using an appropriately configured AWS IAM OpenID Connect (OIDC) identity provider, combined with AWS IAM roles. {product-title} signs service account tokens that are trusted by AWS IAM, and can be projected into a pod and used for authentication. Tokens are refreshed after one hour.

.STS authentication flow
image::142_OpenShift_credentials_STS_0221.svg[Detailed authentication flow between AWS and the cluster when using AWS STS]

> Using manual mode with STS changes the content of the AWS credentials that are provided to individual OpenShift components.

Reviewer 2: s/OpenShift/{product-title}

.AWS secret format using long-lived credentials

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  namespace: <target-namespace> <1>
  name: <target-secret-name> <2>
data:
  aws_access_key_id: <base64-encoded-access-key-id>
  aws_secret_access_key: <base64-encoded-secret-access-key>
----
<1> The namespace for the component.

> <2> The secret name for the component.

Reviewer 2: Should this be _The name of the component secret_ or similar? 
(This makes is seem like the component itself has a name no one should know! Ssh! The name is a secret!)

Reviewer 2: Do we need to describe what the secret is used for?

.AWS secret format with STS

[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  namespace: <target-namespace> <1>
  name: <target-secret-name> <2>
data:
  role_name: <operator-role-name> <3>
  web_identity_token_file: <path-to-token> <4>
----
<1> The namespace for the component.
<2> The secret name for the component.
<3> The IAM role for the component.
<4> The path to the service account token inside the pod. By convention, this is `/var/run/secrets/openshift/serviceaccount/token` for {product-title} components.

[id="sts-mode-installing"]

> == Installing an {product-title} cluster configured for manual mode with STS

Reviewer 2: Should these be separate modules in the assembly?
```
Using manual mode with STS
* Installing with pre-flight
* Installing manually
* Verifying
```

Reviewer 2: I should think the text below can stay in the assemble allowing you to link to the modules.

Author: I was under the impression that I shouldn't create modules unless the content might be used elsewhere - in this case I can't think of another use for it.

In {product-title} version 4.7, you can install a cluster that is configured to use the CCO in manual mode with STS by using one of two methods:

* xref:#sts-mode-installing-preflight-config[Configuring a cluster with the `sts-preflight` tool]
* xref:#sts-mode-installing-manual-config[Configuring a cluster manually]

After installing with either of these methods, the procedure for xref:#sts-mode-installing-verifying[verifying the installation] is the same.

[id="sts-mode-installing-preflight-config"]
=== Configuring a cluster with the sts-preflight tool

When xref:#sts-mode-installing[installing] an {product-title} cluster that is configured to use the CCO in manual mode with STS, you can use the `sts-preflight` tool to create the required AWS resources.

> This tool, which is hosted in an external repository, is provided to help configure a cluster that uses STS mode while the feature is a link:https://access.redhat.com/support/offerings/techpreview[Technology Preview].

Reviewer 2: Do we need _is provided to_? 
_This tool... helps configure a cluster..._

_while the feature is a_ This sounds like the tool will go away at GA. Is this the case?

Reviewer: The tool is available as long as github stays up and running, or as long as the developer doesn't archive/delete the repo. But long-term, we are developing a different tool to handle what sts-preflight does.

+
[source,terminal]
----
$ ./openshift-install version
----
+
.Example output
[source,terminal]
----
./openshift-install 4.7.0
built from commit 2d8b61dd8802058ffffcdc576e0cff67acebdb85
release image registry.svc.ci.openshift.org/ocp/release@sha256:a322ab4941701452f428f5c138efb5354ad2aec33f6f7693f0a3033bfe257d8c
----

. Extract the AWS `CredentialsRequest` custom resources (CRs) from the release image:
+
[source,terminal]
----
$ mkdir credreqs ; oc adm release extract --cloud=aws --credentials-requests quay.io/openshift-release-dev/ocp-release:4.<y>.<z>-x86_64 --to=./credreqs ; cat ./credreqs/*.yaml > credreqs.yaml
----
+
Where `<y>` and `<z>` are the numbers corresponding to the version of {product-title} you are installing.

. Create the `install-config.yaml` file:
+
[source,terminal]
----
$ ./openshift-install create install-config
----

. Configure the cluster to install with the CCO in manual mode:
+
[source,terminal]
----
$ echo "credentialsMode: Manual" >> install-config.yaml
----

. Create install manifests:
+
[source,terminal]
----
$ ./openshift-install create manifests
----

[id="sts-mode-installing-preflight-using"]
==== Using the sts-preflight tool

.Prerequisites

> * Download the link:https://github.com/sjenning/sts-preflight[`sts-preflight` tool].

Reviewer 2: Download to any specific directory or anywhere? 
(I wonder if downloading and running in the install directory would put the output files where they need to go, rather than copying them?)

Author: Another great question. Seems it creates more than just the manifests the user needs to copy over though, so maybe not?

Reviewer: It's really download *and build* the sts-preflight tool. Once the tool is built, you can move it/run it from wherever.

.Procedure

. Using an AWS account with administrator-level permissions, run the following command to:
+
--
** Generate public and private `ServiceAccount` object signing keys.
** Create an S3 bucket.
** Upload the OIDC config into the S3 bucket.
** Set up an IAM Identity Provider that trusts the S3 bucket.
** Create an IAM role with administrator rights to use for installation.
** Create IAM roles for each AWS `CredentialsRequest` CR.
+
[source,terminal]
----
$ ./sts-preflight create --infra-name <aws-infra-name> --region <aws-region> --credentials-requests-to-roles <path-to-credreqs.yaml>
----
+

> Where `<aws-infra-name>` is a prefix the `sts-preflight` tool uses when naming the objects that are created in AWS.

Reviewer 2: AWS folks likely know what the `aws-infra-name` is. Should we mention how to find it anyway?

Author: From what I understood, it's just a generated prefix. So later when they see it in e.g. `<aws-infra-name>-openshift-image-registry-installer-cloud-credentials` they would recognize "oh, it's that prefix"

--
+
The files that are required by the installer are added to the `_output` directory.

> . The `sts-preflight` tool does not put conditions on the IAM roles it creates to allow only specific `ServiceAccount` objects to assume a role. To restrict which identities can assume which role, you must manually modify the trust relationship of each role to add a corresponding condition.

Reviewer 2: _tool does not put conditions on the IAM roles it creates to allow only specific `ServiceAccount` objects to assume a role_
Not clear to me what's going on here. Is this what we want or a negative consequence? 
_tool does not put conditions on the IAM roles it creates, **which allows** only specific `ServiceAccount` objects to assume a role_ ??

Author: Negative consequence. I flipped it around and added a complementizer - still a tricky sentence but I hope more clear?
_The `sts-preflight` tool does not put conditions that allow only specific `ServiceAccount` objects to assume a role on the IAM roles that it creates._

+
[NOTE]
====
Role names that exceed 64 characters might be truncated.
====

> .. Modify the trust relationship of the `<aws-infra-name>-openshift-image-registry-installer-cloud-credentials` role to have the following condition:

Reviewer 2: Will folks know where this and the following 3 parameters are? (I don't!)

Author: Good question for @joelddiaz :) Is this obvious to the user that would be doing this, or does it need some context?

Reviewer: I don't believe we want to reproduce the excellent documentation that AWS already provides about IAM Roles. I would presume that any user interested in running a cluster this way already has an advanced level of understanding of why running a cluster this way is desired, and should be familiar with something as fundamental as IAM Roles.

Author: Ok, good to know! I'm not always certain of the capabilities and knowledge of our audience yet.

> I don't believe we want to reproduce the excellent documentation that AWS already provides

And every tech writer said "Amen" :)

      "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-controller-sa"
    ]
  }
}
----

.. Modify the trust relationship of the `<aws-infra-name>-openshift-machine-api-aws-cloud-credentials` role to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {
    "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [
      "system:serviceaccount:openshift-machine-api:machine-api-controllers"
    ]
  }
}
----

. Copy the manifests that the `sts-preflight` tool created to the `manifests` directory in the same location as the `install-config.yaml` file:
+
[source,terminal]
----
$ cp _output/manifests/* <path-to-install-config.yaml-dir>/manifests/
----

. Copy the private key for the `ServiceAccount` object signer to the same location as the `install-config.yaml` file:
+
[source,terminal]
----
$ cp -a _output/tls <path-to-install-config.yaml-dir>
----

. Create a token that is trusted by the OIDC identity provider:
+
[source,terminal]
----
$ ./sts-preflight token
----
+
This token is used to assume a role to get the credentials that are required to start the installation. The token is created in the `_output/token` directory.

> . Use the token to assume the installer role and mint the credentials that are required for the {product-title} installation process.

Reviewer 2: Will folks know what _mint the credentials_ means? All I see on Google is links to Mint Mode docs and Linux Mint.

Author: s/mint/create :)

----
+
[source,terminal]
----
$ export AWS_SECRET_ACCESS_KEY=<aws-secret-access_>
----
+
[source,terminal]
----
$ export AWS_SESSION_TOKEN=<aws-session-token>
----

[id="sts-mode-installing-preflight-run-installer"]
==== Running the installer

. Run the {product-title} installer
+
[source,terminal]
----
$ ./openshift-install create cluster --log-level=debug
----

. Continue to xref:#sts-mode-installing-verifying[Verifying the installation].

[id="sts-mode-installing-manual-config"]
=== Configuring a cluster manually

When xref:#sts-mode-installing[installing] an {product-title} cluster that is configured to use the CCO in manual mode with STS, you can manually create the required AWS resources.

As an alternative, you can use the xref:#sts-mode-installing-preflight-config[`sts-preflight` tool] to create AWS resources.

.Procedure

. Generate a private key to sign the `ServiceAccount` object:
+
[source,terminal]
----
$ openssl genrsa -out sa-signer 4096
----

. Generate a `ServiceAccount` object public key:
+
[source,terminal]
----
$ openssl rsa -in sa-signer -pubout -out sa-signer.pub
----

. Create an S3 bucket to hold the OIDC configuration:
+
[source,terminal]
----

> $ aws s3api create-bucket --bucket <oidc-bucket-name> --region <aws-region>

Reviewer 2: My ignorance: Can the bucket be named anything or is there a specific name for the bucket already?

Author: It looks like [Amazon has some rules for this](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html), but I am guessing that this would be out of scope for our docs (and also I suspect someone doing this process is familiar with said rules)

Reviewer: Yes, plenty of rules around naming of buckets. This step should error in obvious ways if you pick an invalid bucket name.

----

. Retain the S3 bucket URL:
+

> [source,terminal]

Reviewer 2: Is this a template of the output? The URL I should retain?

Author: I think the latter - @joelddiaz are we setting an env variable?

Reviewer: These steps are not something you can just copy paste. There is a certain amount of reading between the lines implied in many places.
This was an attempt to build a detailed summary of the important pieces of information needed to allow someone to perform every step manually (instead of running the sts-preflight tool).

----
OPENID_BUCKET_URL="https://s3.<aws-region>.amazonaws.com/<oidc-bucket-name>
----

. Build an OIDC configuration:

> .. Save a file named `keys.json` that contains the following information:

Reviewer 2: s/Save/Create ??

+
[source,json]
----
{
    "keys": [
        {
            "use": "sig",
            "kty": "RSA",
            "kid": "<public-signing-key-id>",
            "alg": "RS256",
            "n": "<public-signing-key-modulus>",
            "e": "<public-signing-key-exponent>"
        }
    ]
}
----
+
Where:

*** `<public-signing-key-id>` is generated from the public key with:
+
[source,terminal]
----
$ openssl rsa -in sa-signer.pub -pubin --outform DER | openssl dgst -binary -sha256 | openssl base64 | tr '/+' '_-' | tr -d '='
----
+

> This command converts the public key to DER format, performs a SHA-256 checksum on the binary representation, encodes the data with base64 encoding, and then changes base64-encoded output to use only characters that can be used in a URL.

Reviewer 2: s/and then changes base64-encoded output/and then changes the base64-encoded output... ??

Or;
and changes the base64-encoded output... ??

Reviewer 2: Same in all 3 bullets here

*** `<public-signing-key-modulus>` is generated from the public key with:
+
[source,terminal]
----
$ openssl rsa -pubin -in sa-signer.pub -modulus -noout | sed  -e 's/Modulus=//' | xxd -r -p | base64 -w0 | tr '/+' '_-' | tr -d '='
----
+

> This command prints the Modulus of the public key, extracts the hex representation of the Modulus, converts the ASCII hex to binary, encodes the data with base64 encoding, and then changes base64-encoded output to use only characters that can be used in a URL.

Reviewer 2: Should _Modulus_ be capped? Inconsistent on a Google search (even some with all caps!). 
FWIW IBM seems to use lower-case M:
https://www.ibm.com/support/knowledgecenter/SSLTBW_2.1.0/com.ibm.zos.v2r1.csfb400/csfb4za2578.htm

Author: I'll defer to IBM unless otherwise instructed :)

*** `<public-signing-key-exponent>` is generated from the public key with:
+
[source,terminal]
----
$ printf "%016x" $(openssl rsa -pubin -in sa-signer.pub -noout -text | grep Exponent | awk '{ print $2 }') |  awk '{ sub(/(00)+/, "", $1); print $1 }' | xxd -r -p | base64 -w0 | tr '/+' '_-' | tr -d '='
----
+

> This command extracts decimal representation of public key exponent, prints it as hex with a padded `0` if needed, removes leading `00` pairs, converts the ASCII hex to binary, encodes the data with base64 encoding, and then changes base64-encoded output to use only characters that can be used in a URL.

Reviewer 2: s/extracts decimal representation of public key exponent/extracts **the** decimal representation of **the** public key exponent ??

> .. Save a file named `openid-configuration` that contains the following information:

Reviewer 2: s/Save/Create ??

+
[source,json]
----
{
	"issuer": "$OPENID_BUCKET_URL",
	"jwks_uri": "${OPENID_BUCKET_URL}/keys.json",
    "response_types_supported": [
        "id_token"
    ],
    "subject_types_supported": [
        "public"
    ],
    "id_token_signing_alg_values_supported": [
        "RS256"
    ],
    "claims_supported": [
        "aud",
        "exp",
        "sub",
        "iat",
        "iss",
        "sub"
    ]
}
----

. Upload the OIDC configuration:
+
[source,terminal]
----

> $ aws s3api put-object --bucket <oidc-bucket-name> --key keys.json --body ./keys.json

Reviewer 2: This is the bucket I created earlier?

----
+
[source,terminal]
----
$ aws s3api put-object --bucket <oidc-bucket-name> --key '.well-known/openid-configuration' --body ./openid-configuration
----

. Allow the AWS IAM OIDC identity provider to read these files:
+
[source,terminal]
----
$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key keys.json --acl public-read
----
+
[source,terminal]
----
$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key '.well-known/openid-configuration' --acl public-read
----

. Create an AWS IAM OIDC identity provider:

.. Get the certificate chain from the server that hosts the OIDC configuration:
+
[source,terminal]
----
$ echo | openssl s_client -servername s3.$<aws-region>.amazonaws.com -connect s3.$<aws-region>.amazonaws.com:443 -showcerts 2>/dev/null | awk '/BEGIN/,/END/{ if(/BEGIN/){a++}; out="cert"a".pem"; print >out}'
----

.. Calculate the fingerprint for the certificate at the root of the chain:
+
[source,terminal]
----
$ export BUCKET_FINGERPRINT=$(openssl x509 -in cert<number>.pem -fingerprint -noout | sed -e 's/.*Fingerprint=//' -e 's/://g')
----
+

> Where `<number>` is the highest number in the files that were saved. For example, if two is the highest number in the files that were saved, use `cert2.pem`.

Reviewer 2: s/if two is the highest number/if `2` is the highest number ??

.. Create the identity provider:
+
[source,terminal]
----
$ aws iam create-open-id-connect-provider --url $OPENID_BUCKET_URL --thumbprint-list $BUCKET_FINGERPRINT
----

> .. Save the returned ARN of the newly created identity provider to `$AWS_IAM_OPENID_ARN`.

Reviewer 2: My ignorance: Not sure how to do this step.

Author: I think this is another env variable? @joelddiaz

Reviewer: Again, implied steps. Just telling someone to keep track of this details as it will be needed in a future step.

Author: To attempt to make this clearer but not lose people I have:
> Retain the returned ARN of the newly created identity provider. This ARN is later referred to as `<aws-iam-openid-arn>`.

And then where is it used:
```
[...]
{
                    "Effect": "Allow",
                    "Principal": {
                        "Federated": "<aws-iam-openid-arn>"
                    },
[...]
```
> Where `<aws-iam-openid-arn>` is the returned ARN of the newly created identity provider.

Hopefully that keeps anyone from getting lost!

. Generate IAM roles:

.. Locate all `CredentialsRequest` CRs in this release image that target the cloud you are deploying on:
+
[source,terminal]
----

> $ oc adm release extract quay.io/openshift-release-dev/ocp-release:4.y.z-x86_64 --credentials-requests --cloud=aws

Reviewer 2: Do we need the _Where `<y>` and `<z>` are the numbers corresponding to the version of {product-title} you are installing._ statement here, too?

----

.. For each `CredentialsRequest` CR, create an IAM role that grants the necessary permissions and establishes a trust relationship that trusts the  identity provider previously created.
+
For example, for the openshift-machine-api-operator `CredentialsRequest` CR in `0000_30_machine-api-operator_00_credentials-request.yaml`, create an IAM role that allows an identity from the created OIDC provider created for the cluster:
+
[source,terminal]
----

> $ aws iam get-role --role-name openshift-machine-api-aws-cloud-credentials

Reviewer 2: Does this command create the role or just display it? See: https://docs.aws.amazon.com/cli/latest/reference/iam/get-role.html

Author: @joelddiaz  similarly here, should we tell them to [create-role](https://docs.aws.amazon.com/cli/latest/reference/iam/create-role.html) and to refer to the example for what they need to put into it?

Reviewer: Example output of what the role should look like once the role is created/modified.

----

> +

Reviewer 2: Does this need _.Example output_ or similar to clarify what it is?

[source,json]
----
{
    "Role": {
        "Path": "/",
        "RoleName": "openshift-machine-api-aws-cloud-credentials",
        "RoleId": "ARSOMEROLEID",
        "Arn": "arn:aws:iam::123456789012:role/openshift-machine-api-aws-cloud-credentials",
        "CreateDate": "2021-01-06T15:54:13Z",
        "AssumeRolePolicyDocument": {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {
                        "Federated": "$AWS_IAM_OPENID_ARN"
                    },
                    "Action": "sts:AssumeRoleWithWebIdentity",
                    "Condition": {
                        "StringEquals": {
                            "s3.<aws-region>.amazonaws.com/$BUCKET_NAME:aud": "openshift"
                        }
                    }
                }
            ]
        },
        "Description": "OpenShift role for openshift-machine-api/aws-cloud-credentials",
        "MaxSessionDuration": 3600,
        "RoleLastUsed": {
            "LastUsedDate": "2021-02-03T02:51:24Z",
            "Region": "<aws-region>"
        }
    }
}
----

> .. To further restrict the role such that only specific cluster `ServiceAccount` objects can assume the role, update the `.Role.AssumeRolePolicyDocument.Statement[].Condition` field to the specific `ServiceAccount` object for each component.

Reviewer 2: Are these steps similar to Step 2 in pre-flight? I like the _modify the trust relationship_ wording up there.

Author: I think that's right - updated to make the construction the same.

Author: Changed to make more similar

      "system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"
    ]
  }
}
----

*** For `openshift-ingress-operator` it should contain:
+
[source,json]
----
"Condition": {
  "StringEquals": {
    "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [
      "system:serviceaccount:openshift-ingress-operator:ingress-operator"
    ]
  }
}
----

*** For `openshift-cluster-csi-drivers` it should contain:
+
[source,json]
----
"Condition": {
  "StringEquals": {
    "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [
      "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-operator",
      "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-controller-sa"
    ]
  }
}
----

*** For `openshift-machine-api` it should contain:
+
[source,json]
----
"Condition": {
  "StringEquals": {
    "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [
      "system:serviceaccount:openshift-machine-api:machine-api-controllers"
    ]
  }
}
----

. For each IAM role, attach an IAM policy to the role that reflects the required permissions. For example, for `openshift-machine-api`:
+
[source,terminal]
----

> $ aws iam get-role-policy --role-name openshift-machine-api-aws-cloud-credentials --policy-name openshift-machine-api-aws-cloud-credentials

Reviewer 2: Does this need _.Example output_ or similar to clarify what it is?

> $ aws iam get-role-policy --role-name openshift-machine-api-aws-cloud-credentials --policy-name openshift-machine-api-aws-cloud-credentials

Reviewer 2: Does this command do the attaching or just display the policy? See: https://docs.aws.amazon.com/cli/latest/reference/iam/get-role-policy.html

Author: I see what you mean. @joelddiaz is this just showing them what the policy they attach should look like? And if so, should we tell them to use [attach-role-policy](https://docs.aws.amazon.com/cli/latest/reference/iam/attach-role-policy.html) and to refer to the example for what they need to put into it?

Reviewer: This is an example of what the policy should look like after the policy has been created/applied..

----
+

> [source,json]

Reviewer 2: Does this need _.Example output_ or similar to clarify what it is?

                    "kms:Encrypt",
                    "kms:GenerateDataKey",
                    "kms:GenerateDataKeyWithoutPlainText",
                    "kms:DescribeKey"
                ],
                "Resource": "*"
            },
            {
                "Effect": "Allow",
                "Action": [
                    "kms:RevokeGrant",
                    "kms:CreateGrant",
                    "kms:ListGrants"
                ],
                "Resource": "*",
                "Condition": {
                    "Bool": {
                        "kms:GrantIsForAWSResource": true
                    }
                }
            }
        ]
    }
}
----
. Prepare to run the {product-title} installer:

.. Create the `install-config.yaml` file:
+
[source,terminal]
----
$ ./openshift-install create install-config
----

.. Configure the cluster to install with the CCO in manual mode:
+
[source,terminal]
----
$ echo "credentialsMode: Manual" >> install-config.yaml
----

.. Create install manifests:
+
[source,terminal]
----
$ ./openshift-install create manifests
----

.. Create a `tls` directory, and copy the private key generated previously there:
+
[NOTE]
====

> The target file name is important.

Reviewer 2: s/is important/must be `./tls/bound-service-account-signing-key.key` ??

====
+
[source,terminal]
----
$ mkdir tls ; cp <path-to-service-account-signer> ./tls/bound-service-account-signing-key.key
----

> .. Create a custom `Authentication.config.openshift.io` CR with the file name `cluster-authentication-02-config.yaml`:

Reviewer 2: `Authentication.config.openshift.io` CR
Is this a convention to include the API in a CR title?

+
[source,terminal]
----
$ cat << EOF > manifests/cluster-authentication-02-config.yaml

> iVersion: config.openshift.io/v1

Reviewer 2: s/iVersion/apiVersion (I do this all the time when pasting into a Linux file!)

kind: Authentication
metadata:
  name: cluster
spec:
  serviceAccountIssuer: $OPENID_BUCKET_URL
EOF
----

> .. For each `CredentialsRequest` CR that is extracted from the release image, create a secret with the target namespace and target name that is indicated in each `CredentialsRequest`, substituting the AWS IAM role ARN created previously for each component:

Reviewer 2: Do we need to add the `oc adm release extract` command again as a reminder?

Author: It can't hurt!

+
.Example secret file for `openshift-machine-api`:
+
[source,terminal]
----
$ cat manifests/openshift-machine-api-aws-cloud-credentials-credentials.yaml
apiVersion: v1
stringData:
  credentials: |-
    [default]
    role_arn = arn:aws:iam::123456789012:role/openshift-machine-api-aws-cloud-credentials
    web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
kind: Secret
metadata:
  name: aws-cloud-credentials
  namespace: openshift-machine-api
type: Opaque
----

[id="sts-mode-installing-manual-run-installer"]
==== Running the installer

. Run the {product-title} installer:
+
[source,terminal]
----
$ ./openshift-install create cluster
----

. Continue to xref:#sts-mode-installing-verifying[Verifying the installation].

[id="sts-mode-installing-verifying"]
=== Verifying the installation

. Connect to the {product-title} cluster.

. Verify that the cluster does not have `root` credentials:
+
[source,terminal]
----
$ oc get secrets -n kube-system aws-creds
----
+
The output should look similar to:
+
[source,terminal]
----
Error from server (NotFound): secrets "aws-creds" not found
----

> . Verify that the components are assuming the IAM roles that are specified in the secret manifests, instead of using credentials that are minted by the CCO:

Reviewer 2: What is a _secret manifest_? Seems to be the first reference to that term.

Author: afaict, yes. I think manifest is probably a better term?

[id="sts-mode-installing-preflight-config"]
=== Configuring a cluster with the sts-preflight tool

When xref:#sts-mode-installing[installing] an {product-title} cluster that is configured to use the CCO in manual mode with STS, you can use the `sts-preflight` tool to create the required AWS resources.

This tool, which is hosted in an external repository, is provided to help configure a cluster that uses STS mode while the feature is a link:https://access.redhat.com/support/offerings/techpreview[Technology Preview].

As an alternative, you can xref:#sts-mode-installing-manual-config[create the required AWS resources manually].

[id="sts-mode-installing-preflight-preparing"]
==== Preparing the installer

.Prerequisites
* Obtain the pull secret for your cluster.
* Obtain the {product-title} version 4.7 installation binary.

.Procedure

. From the directory that contains the installation program, obtain details of the {product-title} release image that your `openshift-install` binary is built to use:
+
[source,terminal]
----
$ openshift-install version
----
+
.Example output
[source,terminal]
----

> release image quay.io/openshift-release-dev/ocp-release:4.y.z-x86_64

Author: I snagged that from the Manually Creating IAM content - is it incorrect there as well?

> release image quay.io/openshift-release-dev/ocp-release:4.y.z-x86_64

Reviewer: The formatting is correct. The sha256 is the unique identifier for any release image, and then we tag it with human-readable strings. But the installer only gives the sha256 hash (not the human-readable version tag). You have to go digging around quay.io to see what tag(s) might be associated with any given sha256 hash.

----

. Extract the AWS `CredentialsRequest` custom resources (CRs) from the release image:
+
[source,terminal]
----
$ mkdir credreqs ; oc adm release extract --cloud=aws --credentials-requests quay.io/openshift-release-dev/ocp-release:4.y.z-x86_64 --to=./credreqs ; cat ./credreqs/*.yaml > credreqs.yaml
----

. Create the `install-config.yaml` file:
+
[source,terminal]
----
$ ./openshift-install create install-config
----

. Configure the cluster to install with the CCO in manual mode:
+
[source,terminal]
----
$ echo "credentialsMode: Manual" >> install-config.yaml
----

. Create install manifests:
+
[source,terminal]
----
$ ./openshift-install create manifests
----

[id="sts-mode-installing-preflight-using"]
==== Using the sts-preflight tool

.Prerequisites
* Download the link:https://github.com/sjenning/sts-preflight[`sts-preflight` tool].

> * Ensure that you have an AWS account with sufficient permissions.

Author: In this case, I'll just have it in the step rather than as a pre-req :)

+
[source,terminal]
----
$ ./sts-preflight assume
----
+
The output prints several `export` commands.

. Run the `export` commands from the output of the previous step.
+
[source,terminal]
----
$ export AWS_ACCESS_KEY_ID=<aws-access-key-id>
----
+
[source,terminal]
----
$ export AWS_SECRET_ACCESS_KEY=<aws-secret-access_>
----
+
[source,terminal]
----
$ export AWS_SESSION_TOKEN=<aws-session-token>
----

[id="sts-mode-installing-preflight-run-installer"]
==== Running the installer

. Run the {product-title} installer
+
[source,terminal]
----
$ ./openshift-install create cluster --log-level=debug
----

. Continue to xref:#sts-mode-installing-verifying[Verifying the installation].

[id="sts-mode-installing-manual-config"]
=== Configuring a cluster manually

When xref:#sts-mode-installing[installing] an {product-title} cluster that is configured to use the CCO in manual mode with STS, you can manually create the required AWS resources.

As an alternative, you can use the xref:#sts-mode-installing-preflight-config[`sts-preflight` tool] to create AWS resources.

> .Procedure

Reviewer: FYI: Still unverified procedure...

----

. Generate a `ServiceAccount` object public key:
+
[source,terminal]
----
$ openssl rsa -in sa-signer -pubout -out sa-signer.pub
----

. Create an S3 bucket to hold the OIDC configuration:
+
[source,terminal]
----
$ aws s3api create-bucket --bucket <oidc-bucket-name> --region <aws-region>
----

. Retain the S3 bucket URL:
+
[source,terminal]
----
OPENID_BUCKET_URL="https://s3.<aws-region>.amazonaws.com/<oidc-bucket-name>
----

. Build an OIDC configuration:

.. Save a file named `keys.json` that contains the following information:
+
[source,json]
----
{
    "keys": [
        {
            "use": "sig",
            "kty": "RSA",
            "kid": "<public-signing-key-id>",
            "alg": "RS256",
            "n": "<public-signing-key-modulus>",
            "e": "<public-signing-key-exponent>"
        }
    ]
}
----
+
Where:

*** `<public-signing-key-id>` is generated from the public key with:
+
[source,terminal]
----
$ openssl rsa -in sa-signer.pub -pubin --outform DER | openssl dgst -binary -sha256 | openssl base64 | tr '/+' '_-' | tr -d '='
----
+
This command converts the public key to DER format, performs a SHA-256 checksum on the binary representation, and
+

> <needinfo> What's it doing with this part `base64 encode | base64url encode`?

Reviewer: It's two passes of changes to the output. The first is to base64 encode the data, but base64 encoding allows certain characters that aren't allowed to appear in URLs, so they created "base64 url" encoding to make the changes necessary to the raw base64 encoded data so that the "base64 url" encoded data can be used in URLs.

Author: Neat! Thanks :)

                    },
                    "Action": "sts:AssumeRoleWithWebIdentity",
                    "Condition": {
                        "StringEquals": {
                            "s3.<aws-region>.amazonaws.com/$BUCKET_NAME:aud": "openshift"
                        }
                    }
                }
            ]
        },
        "Description": "OpenShift role for openshift-machine-api/aws-cloud-credentials",
        "MaxSessionDuration": 3600,
        "RoleLastUsed": {
            "LastUsedDate": "2021-02-03T02:51:24Z",
            "Region": "<aws-region>"
        }
    }
}
----
+
Where `<aws-iam-openid-arn>` is the returned ARN of the newly created identity provider.

.. To further restrict the role such that only specific cluster `ServiceAccount` objects can assume the role, modify the trust relationship of each role by updating the `.Role.AssumeRolePolicyDocument.Statement[].Condition` field to the specific `ServiceAccount` object for each component.

*** Modify the trust relationship of the `cluster-image-registry-operator` role to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: @jeana-redhat  This line changing to `"s3.<aws-region>.amazonaws.com/$BUCKET_NAME:sub": [` would be better to make context consistent with line 482.

      "system:serviceaccount:openshift-image-registry:registry",
      "system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"
    ]
  }
}
----

*** Modify the trust relationship of the `openshift-ingress-operator` to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: changing to `"s3.<aws-region>.amazonaws.com/$BUCKET_NAME:sub": [`

      "system:serviceaccount:openshift-ingress-operator:ingress-operator"
    ]
  }
}
----

*** Modify the trust relationship of the `openshift-cluster-csi-drivers` to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: changing to `"s3.<aws-region>.amazonaws.com/$BUCKET_NAME:sub": [`

      "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-operator",
      "system:serviceaccount:openshift-cluster-csi-drivers:aws-ebs-csi-driver-controller-sa"
    ]
  }
}
----

*** Modify the trust relationship of the `openshift-machine-api` to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: changing to `"s3.<aws-region>.amazonaws.com/$BUCKET_NAME:sub": [`

** Upload the OIDC config into the S3 bucket.
** Set up an IAM Identity Provider that trusts the S3 bucket.
** Create an IAM role with administrator rights to use for installation.
** Create IAM roles for each AWS `CredentialsRequest` CR.
+
[source,terminal]
----
$ ./sts-preflight create --infra-name <aws-infra-name> --region <aws-region> --credentials-requests-to-roles <path-to-credreqs.yaml>
----
+
Where `<aws-infra-name>` is a prefix the `sts-preflight` tool uses when naming the objects that are created in AWS.
--
+
The files that are required by the installer are added to the `_output` directory.

. The `sts-preflight` tool does not put conditions that allow only specific `ServiceAccount` objects to assume a role on the IAM roles that it creates. To restrict which identities can assume which role, you must manually modify the trust relationship of each role to add a corresponding condition.
+
[NOTE]
====
Role names that exceed 64 characters might be truncated.
====

.. Modify the trust relationship of the `<aws-infra-name>-openshift-image-registry-installer-cloud-credentials` role to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: It would be good to change this more rigorous : `"s3.<aws-region>.amazonaws.com/<aws-infra-name>-installer:sub"`

      "system:serviceaccount:openshift-image-registry:registry",
      "system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"
    ]
  }
}
----

.. Modify the trust relationship of the `<aws-infra-name>-openshift-ingress-operator-cloud-credentials` role to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: the same as above: `"s3.<aws-region>.amazonaws.com/<aws-infra-name>-installer:sub"`

      "system:serviceaccount:openshift-ingress-operator:ingress-operator"
    ]
  }
}
----

.. Modify the trust relationship of the `<aws-infra-name>-openshift-cluster-csi-drivers-ebs-cloud-credentials` role to have the following condition:
+
[source,json]
----
"Condition": {
  "StringEquals": {

>     "s3.us-east-1.amazonaws.com/<aws-infra-name>-installer:sub": [

Reviewer 3: the same as above: `"s3.<aws-region>.amazonaws.com/<aws-infra-name>-installer:sub"`

----

[id="sts-mode-installing-preflight-run-installer"]
==== Running the installer

. Run the {product-title} installer
+
[source,terminal]
----
$ ./openshift-install create cluster --log-level=debug
----

. Continue to xref:#sts-mode-installing-verifying[Verifying the installation].

[id="sts-mode-installing-manual-config"]
=== Configuring a cluster manually

When xref:#sts-mode-installing[installing] an {product-title} cluster that is configured to use the CCO in manual mode with STS, you can manually create the required AWS resources.

As an alternative, you can use the xref:#sts-mode-installing-preflight-config[`sts-preflight` tool] to create AWS resources.

.Procedure

. Generate a private key to sign the `ServiceAccount` object:
+
[source,terminal]
----
$ openssl genrsa -out sa-signer 4096
----

. Generate a `ServiceAccount` object public key:
+
[source,terminal]
----
$ openssl rsa -in sa-signer -pubout -out sa-signer.pub
----

. Create an S3 bucket to hold the OIDC configuration:
+
[source,terminal]
----
$ aws s3api create-bucket --bucket <oidc-bucket-name> --region <aws-region>
----

. Retain the S3 bucket URL:
+
[source,terminal]
----

> OPENID_BUCKET_URL="https://s3.<aws-region>.amazonaws.com/<oidc-bucket-name>

Reviewer 3: Missing `"` in the end

. Allow the AWS IAM OIDC identity provider to read these files:
+
[source,terminal]
----
$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key keys.json --acl public-read
----
+
[source,terminal]
----
$ aws s3api put-object-acl --bucket <oidc-bucket-name> --key '.well-known/openid-configuration' --acl public-read
----

. Create an AWS IAM OIDC identity provider:

.. Get the certificate chain from the server that hosts the OIDC configuration:
+
[source,terminal]
----
$ echo | openssl s_client -servername s3.$<aws-region>.amazonaws.com -connect s3.$<aws-region>.amazonaws.com:443 -showcerts 2>/dev/null | awk '/BEGIN/,/END/{ if(/BEGIN/){a++}; out="cert"a".pem"; print >out}'
----

.. Calculate the fingerprint for the certificate at the root of the chain:
+
[source,terminal]
----
$ export BUCKET_FINGERPRINT=$(openssl x509 -in cert<number>.pem -fingerprint -noout | sed -e 's/.*Fingerprint=//' -e 's/://g')
----
+
Where `<number>` is the highest number in the files that were saved. For example, if `2` is the highest number in the files that were saved, use `cert2.pem`.

.. Create the identity provider:
+
[source,terminal]
----

> $ aws iam create-open-id-connect-provider --url $OPENID_BUCKET_URL --thumbprint-list $BUCKET_FINGERPRINT

Reviewer 3: when creating the identity provider , should be with` --client-id-list openshift`, otherwise the installation will fail.
like:
`$ aws iam create-open-id-connect-provider --url $OPENID_BUCKET_URL --thumbprint-list $BUCKET_FINGERPRINT --client-id-list openshift`

+
For example, for the openshift-machine-api-operator `CredentialsRequest` CR in `0000_30_machine-api-operator_00_credentials-request.yaml`, create an IAM role that allows an identity from the created OIDC provider created for the cluster, similar to the following:
+
[source,json]
----
{
    "Role": {
        "Path": "/",
        "RoleName": "openshift-machine-api-aws-cloud-credentials",
        "RoleId": "ARSOMEROLEID",
        "Arn": "arn:aws:iam::123456789012:role/openshift-machine-api-aws-cloud-credentials",
        "CreateDate": "2021-01-06T15:54:13Z",
        "AssumeRolePolicyDocument": {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {
                        "Federated": "<aws-iam-openid-arn>"
                    },
                    "Action": "sts:AssumeRoleWithWebIdentity",
                    "Condition": {
                        "StringEquals": {

>                             "oidc-bucket-name>.s3.<aws-region>.amazonaws.com/$BUCKET_NAME:aud": "openshift"

Reviewer: ```suggestion
                            "<oidc-bucket-name>.s3.<aws-region>.amazonaws.com/$BUCKET_NAME:aud": "openshift"
```
